File: apps/api/src/lunaris_api/schemas/bookmarks.py
from datetime import datetime

from pydantic import Field

from ..bookmarks import BookmarkKind
from .base import CamelModel


class BookmarkView(CamelModel):
    """One saved lesson/concept/source, with the display fields captured at save time."""

    kind: BookmarkKind
    course_id: str
    target_id: str
    course_title: str | None = None
    title: str | None = None
    lesson_id: str | None = None
    snippet: str | None = None
    concept_tier: int | None = None
    trust_tier: str | None = None
    credibility: float | None = None
    note: str | None = None
    saved_at: datetime


class BookmarkRequest(CamelModel):
    """Save (idempotent upsert on the natural key). Bounds mirror the DB checks; display fields
    are captured client-side at save time (the screen must render without re-fetching courses)."""

    kind: BookmarkKind
    course_id: str = Field(min_length=1, max_length=100)
    target_id: str = Field(min_length=1, max_length=300)
    course_title: str | None = Field(default=None, min_length=1, max_length=300)
    title: str | None = Field(default=None, min_length=1, max_length=300)
    lesson_id: str | None = Field(default=None, min_length=1, max_length=200)
    snippet: str | None = Field(default=None, min_length=1, max_length=2000)
    concept_tier: int | None = Field(default=None, ge=1, le=5)
    trust_tier: str | None = Field(default=None, min_length=1, max_length=40)
    credibility: float | None = Field(default=None, ge=0, le=1)
    note: str | None = Field(default=None, min_length=1, max_length=2000)

File: apps/web/src/lib/bookmarks.ts
import { authedFetch } from "./apiClient";

export type BookmarkKind = "lesson" | "concept" | "source";

/** The longest claim quote a source bookmark stores — mirrors the DB's snippet check. */
export const MAX_SNIPPET_LENGTH = 2000;

export interface Bookmark {
  kind: BookmarkKind;
  courseId: string;
  targetId: string;
  courseTitle?: string | null;
  title?: string | null;
  lessonId?: string | null;
  snippet?: string | null;
  conceptTier?: number | null;
  trustTier?: string | null;
  credibility?: number | null;
  note?: string | null;
  savedAt: string;
}

/** What a save affordance sends — everything but savedAt (the server stamps it). */
export type BookmarkDraft = Omit<Bookmark, "savedAt">;

/** The toggle's natural key — the client never knows row ids. */
export interface BookmarkRef {
  kind: BookmarkKind;
  courseId: string;
  targetId: string;
}

export class BookmarksError extends Error {
  constructor(message: string, options?: ErrorOptions) {
    super(message, options);
    this.name = "BookmarksError";
  }
}

/** Fetch the caller's saves (newest first). Rejects with BookmarksError on transport/HTTP
 *  failure — and on an alien payload shape (trust-boundary check: consumers read fields
 *  unguarded). */
export async function fetchBookmarks(
  apiBaseUrl: string,
  signal?: AbortSignal,
): Promise<Bookmark[]> {
  let response: Response;
  try {
    response = await authedFetch(`${apiBaseUrl}/api/bookmarks`, signal ? { signal } : undefined);
  } catch (cause) {
    throw new BookmarksError("Could not reach your bookmarks.", { cause });
  }
  if (!response.ok) {
    throw new BookmarksError(`Couldn't load your bookmarks (HTTP ${response.status}).`);
  }
  const body = (await response.json()) as Bookmark[] | null;
  if (!Array.isArray(body) || body.some((item) => typeof item?.targetId !== "string")) {
    throw new BookmarksError("Couldn't read your bookmarks (unexpected response).");
  }
  return body;
}

/** Save (idempotent upsert on the natural key). Rejects on failure so the caller can reconcile
 *  an optimistic toggle. */
export async function putBookmark(apiBaseUrl: string, draft: BookmarkDraft): Promise<void> {
  const response = await authedFetch(`${apiBaseUrl}/api/bookmarks`, {
    method: "PUT",
    headers: { "content-type": "application/json" },
    body: JSON.stringify(draft),
  });
  if (!response.ok)
    throw new BookmarksError(`Couldn't save the bookmark (HTTP ${response.status}).`);
}

/** Remove by the natural key (idempotent). Rejects on failure for optimistic reconcile. */
export async function deleteBookmark(apiBaseUrl: string, ref: BookmarkRef): Promise<void> {
  const params = new URLSearchParams({
    kind: ref.kind,
    courseId: ref.courseId,
    targetId: ref.targetId,
  });
  const response = await authedFetch(`${apiBaseUrl}/api/bookmarks?${params}`, {
    method: "DELETE",
  });
  if (!response.ok) {
    throw new BookmarksError(`Couldn't remove the bookmark (HTTP ${response.status}).`);
  }
}
